### PR TITLE
[Feature] Add Device Flag in Data Loaders

### PR DESCRIPTION
**Pull request by mufeili**
## Description
Add a device flag for `NodeDataLoader` and `EdgeDataLoader`. For `NodeDataLoader`, the device flag controls the device of the generated blocks in each iteration. For `EdgeDataLoader`, the device flag controls the device of the generated blocks and graphs in each iteration.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [ ] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

**Review by jermainewang — APPROVED** [review]
LGTM

**Comment by mufeili** [comment]
Just to be careful, can any of you check the modifications for `dist_dataloader.py`? @zheng-da @aksnzhy @VoVAllen

**Comment by VoVAllen** [comment]
Why we want to add such flag?

**Comment by mufeili** [comment]
> Why we want to add such flag?

I think this just allows users to save additional code like `[blc.to(device) for blc in blocks]` in each iteration.

**Comment by jermainewang** [comment]
> Why we want to add such flag?

Adding this option allows DGL to take control of the data movement, which opens up optimizations such as asynchronous CPU-GPU transfer, prefetching, etc.

**Comment by VoVAllen** [comment]
I'm not sure about the Dataloader API compatibility with DistDataloader. Need @zheng-da to take a look

File: python/dgl/dataloading/pytorch/__init__.py [hunk 1]
```diff
@@ -137,17 +137,20 @@ def collate(self, items):
...
-    def __init__(self, node_dataloader):
+    def __init__(self, device, node_dataloader):
+        self.device = device
         self.node_dataloader = node_dataloader
         self.iter_ = iter(node_dataloader.dataloader)
 
     def __next__(self):
         input_nodes, output_nodes, blocks = next(self.iter_)
         _restore_blocks_storage(blocks, self.node_dataloader.collator.g)
+        blocks = [blc.to(self.device) for blc in blocks]
```
Comment (zheng-da) [inline python/dgl/dataloading/pytorch/__init__.py:148]: does this run in the sampler process or the trainer process?

Comment (mufeili) [reply python/dgl/dataloading/pytorch/__init__.py:148]: @aksnzhy said it's in the sampler process.

Comment (BarclayII) [reply python/dgl/dataloading/pytorch/__init__.py:148]: For single-machine training, this should be in the trainer (or main) process?  The sampler process does not (and should not) return a graph with node/edge features.  The main process is responsible for getting the features.

For distributed training, I'm not sure.  Perhaps the sampler process is responsible for getting both the node/edge features and the graph and sending them to the trainer.

**Review by BarclayII — COMMENTED** [review]
LGTM.

File: python/dgl/dataloading/pytorch/__init__.py [hunk 2]
```diff
@@ -137,17 +137,20 @@ def collate(self, items):
             return input_nodes, pair_graph, neg_pair_graph, blocks
 
 class _NodeDataLoaderIter:
-    def __init__(self, node_dataloader):
+    def __init__(self, device, node_dataloader):
+        self.device = device
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/dataloading/pytorch/__init__.py:141]: Done.

File: python/dgl/dataloading/pytorch/__init__.py [hunk 3]
```diff
@@ -137,17 +137,20 @@ def collate(self, items):
...
         self.node_dataloader = node_dataloader
         self.iter_ = iter(node_dataloader.dataloader)
 
     def __next__(self):
         input_nodes, output_nodes, blocks = next(self.iter_)
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/dataloading/pytorch/__init__.py:146]: Done.

File: python/dgl/dataloading/pytorch/__init__.py [hunk 4]
```diff
@@ -137,17 +137,20 @@ def collate(self, items):
...
         _restore_blocks_storage(blocks, self.node_dataloader.collator.g)
+        blocks = [blc.to(self.device) for blc in blocks]
         return input_nodes, output_nodes, blocks
 
 class _EdgeDataLoaderIter:
-    def __init__(self, edge_dataloader):
+    def __init__(self, device, edge_dataloader):
+        self.device = device
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/dataloading/pytorch/__init__.py:153]: Done.

File: python/dgl/dataloading/pytorch/__init__.py [hunk 5]
```diff
@@ -156,12 +159,17 @@ def __next__(self):
             input_nodes, pair_graph, blocks = next(self.iter_)
             _restore_subgraph_storage(pair_graph, self.edge_dataloader.collator.g)
             _restore_blocks_storage(blocks, self.edge_dataloader.collator.g_sampling)
+            pair_graph = pair_graph.to(self.device)
+            blocks = [blc.to(self.device) for blc in blocks]
             return input_nodes, pair_graph, blocks
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/dataloading/pytorch/__init__.py:164]: Done.

File: python/dgl/dataloading/pytorch/__init__.py [hunk 6]
```diff
@@ -156,12 +159,17 @@ def __next__(self):
...
         else:
             input_nodes, pair_graph, neg_pair_graph, blocks = next(self.iter_)
             _restore_subgraph_storage(pair_graph, self.edge_dataloader.collator.g)
             _restore_subgraph_storage(neg_pair_graph, self.edge_dataloader.collator.g)
             _restore_blocks_storage(blocks, self.edge_dataloader.collator.g_sampling)
+            pair_graph = pair_graph.to(self.device)
+            neg_pair_graph = neg_pair_graph.to(self.device)
+            blocks = [blc.to(self.device) for blc in blocks]
             return input_nodes, pair_graph, neg_pair_graph, blocks
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/dataloading/pytorch/__init__.py:173]: Done.

**Comment by mufeili** [comment]
@BarclayII Can you help resolving the conflicts in `python/dgl/dataloading/pytorch/__init__.py`?

**Pull request merged**